Document markers sometimes get erroneously removed on earlier words when removing text in a later part of the sentence

rr-codes opened the pull request:
#### 8e1031fc9a3ad0d3705c50866a49a019e7370363
<pre>
Document markers sometimes get erroneously removed on earlier words when removing text in a later part of the sentence
<a href="https://bugs.webkit.org/show_bug.cgi?id=268778">https://bugs.webkit.org/show_bug.cgi?id=268778</a>
<a href="https://rdar.apple.com/122336557">rdar://122336557</a>

Reviewed by Megan Gardner and Aditya Keerthi.

When a text node (for example, &quot;AAAA BBBB CCCC&quot;) has a document marker on it (for example, a spelling
marker on &quot;AAAA&quot;), and some text is then removed from it (like &quot;CCCC&quot;), the document marker will get
removed accidentally.

This is because the logic in `shiftMarkers` does not guard against markers that occur in the prior
part of the text, which should just be completely ignored and not removed or shifted, since they
would not be affected by the removal of later text.

Fix by checking for such a case and not removing the marker in that case.

* LayoutTests/editing/spelling/document-markers-remain-after-removing-text.html: Added.
* LayoutTests/platform/ios-simulator/editing/spelling/document-markers-remain-after-removing-text-expected.txt: Added.
* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::DocumentMarkerController::shiftMarkers):

Canonical link: <a href="https://commits.webkit.org/274129@main">https://commits.webkit.org/274129@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/115415633a4b16d121659dd41603d64d984747fc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38014 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16923 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40271 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40560 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33815 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/39767 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19627 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14232 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32119 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38588 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14284 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/33282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12453 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12401 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34007 "Build is being retried. Recent messages:OS: Monterey (12.6.8), Xcode: 14.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed layout tests; 4 flakes") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41835 "Built successfully") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34490 "Build is being retried. Recent messages:OS: Sonoma (14.0), Xcode: 15.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests (warnings)") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38263 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12990 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10635 "Build is being retried. Recent messages:OS: Sonoma (14.1), Xcode: 15.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed layout tests; 4 flakes") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36458 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14551 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/33345 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8523 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13409 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14001 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/platform/ios-simulator/editing/spelling/document-markers-remain-after-removing-text-expected.txt:
```diff
@@ -0,0 +1,9 @@
+      RenderBlock {DIV} at (0,0) size 302x22 [border: (1px solid #000000)]
+        RenderText {#text} at (1,1) size 105x19
+          text run at (1,1) width 101: "aaaa bbbb dddd"
+          text run at (101,1) width 5: " "
```
This result doesn't tell us anything. We either need a reference or some assertions.

pxlcoder replied:
Also, we should avoid platform specific results if possible.

pxlcoder reviewed (commented):
PR title nit: swap "get" and "sometimes"

pxlcoder commented on Source/WebCore/dom/DocumentMarkerController.cpp:
```diff
@@ -644,9 +644,21 @@ void DocumentMarkerController::shiftMarkers(Node& node, unsigned startOffset, in
+        // removed is `startOffset - delta`.
+        auto startOffsetOfAffectedRange = static_cast<int>(startOffset);
+        if (delta < 0)
+            startOffsetOfAffectedRange += delta;
```
I don't think this needs to be computed in the loop?

pxlcoder commented on Source/WebCore/dom/DocumentMarkerController.cpp:
```diff
@@ -644,9 +644,21 @@ void DocumentMarkerController::shiftMarkers(Node& node, unsigned startOffset, in
+        if (delta < 0)
+            startOffsetOfAffectedRange += delta;
+
+        if (marker.endOffset() > clampTo<unsigned>(startOffsetOfAffectedRange)) {
```
Shouldn't this comparison be against `marker.startOffset()`?

rr-codes replied:
I don’t think; this is just testing if the marker overlaps the removed text range; so marker end > removed text range start

pxlcoder commented on LayoutTests/editing/spelling/document-markers-remain-after-removing-text.html:
```diff
@@ -0,0 +1,79 @@
+    text-decoration-line: underline;
+    text-decoration-style: double;
+    text-decoration-color: red;
+}
```
This rule is unused.

pxlcoder commented on LayoutTests/editing/spelling/document-markers-remain-after-removing-text-expected.html:
```diff
@@ -0,0 +1,34 @@
+    text-decoration-line: underline;
+    text-decoration-style: double;
+    text-decoration-color: red;
+}
```
This rule is unused.

pxlcoder commented on Source/WebCore/dom/DocumentMarkerController.cpp:
```diff
@@ -644,13 +644,19 @@ void DocumentMarkerController::shiftMarkers(Node& node, unsigned startOffset, in
         // FIXME: No obvious reason this should be iOS-specific. Remove the #if at some point.
         auto targetStartOffset = clampTo<unsigned>(static_cast<int>(marker.startOffset()) + delta);
         auto targetEndOffset = clampTo<unsigned>(static_cast<int>(marker.endOffset()) + delta);
```
Why are these left behind here?

rr-codes replied:
because they are used in line 667 too

megangardner commented on Source/WebCore/dom/DocumentMarkerController.cpp:
```diff
@@ -644,13 +644,19 @@ void DocumentMarkerController::shiftMarkers(Node& node, unsigned startOffset, in
+            // There is no need to adjust or remove the marker if it's before the start of the
+            // text being removed.
+
+#if PLATFORM(IOS_FAMILY)
```
Why is this only iOSFamly?

rr-codes replied:
because that is how it was before and I didn’t want to risk changing it

megangardner replied:
oh, ok,  the diff cuts off the top where that is stated, I see it now.

rr-codes committed:
Document markers sometimes get erroneously removed on earlier words when removing text in a later part of the sentence
https://bugs.webkit.org/show_bug.cgi?id=268778
rdar://122336557

Reviewed by Megan Gardner and Aditya Keerthi.

When a text node (for example, "AAAA BBBB CCCC") has a document marker on it (for example, a spelling
marker on "AAAA"), and some text is then removed from it (like "CCCC"), the document marker will get
removed accidentally.

This is because the logic in `shiftMarkers` does not guard against markers that occur in the prior
part of the text, which should just be completely ignored and not removed or shifted, since they
would not be affected by the removal of later text.

Fix by checking for such a case and not removing the marker in that case.

* LayoutTests/editing/spelling/document-markers-remain-after-removing-text.html: Added.
* LayoutTests/platform/ios-simulator/editing/spelling/document-markers-remain-after-removing-text-expected.txt: Added.
* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::DocumentMarkerController::shiftMarkers):

Canonical link: https://commits.webkit.org/274129@main
